code/core/write_sdl_ntuple.cc

```diff
@@ -321,131 +252,64 @@ std::tuple<float, float, float, vector<unsigned int>, vector<unsigned int>> pars
     SDL::segments& segmentsInGPU = (*event->getSegments());
     SDL::miniDoublets& miniDoubletsInGPU = (*event->getMiniDoublets());
     SDL::hits& hitsInGPU = (*event->getHits());
-    SDL::modules& modulesInGPU = (*event->getModules());
-    SDL::pixelTriplets& pixelTripletsInGPU = (*event->getPixelTriplets());
-    SDL::objectRanges& rangesInGPU = (*event->getRanges());
-
-    // Main pT3 object idx in TC container
-    unsigned int objectIdx = trackCandidatesInGPU.objectIndices[2 * idx];
-
-    unsigned int pLSIdx = pixelTripletsInGPU.pixelSegmentIndices[objectIdx];
-    unsigned int innerSegmentIdx = tripletsInGPU.segmentIndices[2 * pixelTripletsInGPU.tripletIndices[objectIdx]];
-    unsigned int outerSegmentIdx = tripletsInGPU.segmentIndices[2 * pixelTripletsInGPU.tripletIndices[objectIdx] + 1];
-
-    // Getting MD indices
-    unsigned int pLS_md1 = segmentsInGPU.mdIndices[2 * pLSIdx];
-    unsigned int pLS_md2 = segmentsInGPU.mdIndices[2 * pLSIdx + 1];
-    unsigned int T3_md1 = segmentsInGPU.mdIndices[2 * innerSegmentIdx];
-    unsigned int T3_md2 = segmentsInGPU.mdIndices[2 * innerSegmentIdx + 1];
-    unsigned int T3_md3 = segmentsInGPU.mdIndices[2 * outerSegmentIdx + 1];
-
-    // Getting Hit indices
-    unsigned int pLS_hit1 = miniDoubletsInGPU.anchorHitIndices[pLS_md1];
-    unsigned int pLS_hit2 = miniDoubletsInGPU.outerHitIndices[pLS_md1];
-    unsigned int pLS_hit3 = miniDoubletsInGPU.anchorHitIndices[pLS_md2];
-    unsigned int pLS_hit4 = miniDoubletsInGPU.outerHitIndices[pLS_md2];
-    unsigned int T3_hit1 = miniDoubletsInGPU.anchorHitIndices[T3_md1];
-    unsigned int T3_hit2 = miniDoubletsInGPU.outerHitIndices[T3_md1];
-    unsigned int T3_hit3 = miniDoubletsInGPU.anchorHitIndices[T3_md2];
-    unsigned int T3_hit4 = miniDoubletsInGPU.outerHitIndices[T3_md2];
-    unsigned int T3_hit5 = miniDoubletsInGPU.anchorHitIndices[T3_md3];
-    unsigned int T3_hit6 = miniDoubletsInGPU.outerHitIndices[T3_md3];
 
     // pLS  1    2    3
     // **** o -- o -- o
-    unsigned int i1_idx = T3_hit1;
-    unsigned int i3_idx = T3_hit5;
-    const float dr = sqrt(pow(hitsInGPU.xs[i3_idx] - hitsInGPU.xs[i1_idx], 2) + pow(hitsInGPU.ys[i3_idx] - hitsInGPU.ys[i1_idx], 2));
+    unsigned int pT3 = trackCandidatesInGPU.directObjectIndices[idx];
+    std::vector<unsigned int> Hits = getHitsFrompT3(event, pT3);
+    unsigned int Hit_0 = Hits[0];
+    unsigned int Hit_4 = Hits[4];
```

Comment:
If I read the code correctly, the hit array for pixel hits is of variable length, so it may have either 3 or 4 hits. If that's the case, I am not sure that the 4th hit is always the right choice here (and possibly in other places).

Trying to go back on the function calls:
1. L259 of `write_sdl_ntuple.cc`
2. L229 of `AccessHelper.cc`
3. L233 of `AccessHelper.cc`
4. L8 of `AccessHelper.cc`
5. L21-L24 of `AccessHelper.cc` which gives a variable length array.

If the logic above is indeed what's happening, we would need to understand how to resolve this. I see two ways:
1. We add a "fake" fourth hit, as in the previous logic.
2. We choose the 3rd or 4th hit based on the length of the pLS/pTX object (probably more natural for anything that comes downstream?).

Thinking a bit more about it, maybe we don't care that much, given that in any case the hit we take will be in the same MD, however I find this to be confusing and inconsistent.

Reply:
Ah found this comment now.
This seems like a bug!!!! 
Fixed it with c3f39ff